### PR TITLE
Restore email notifications

### PR DESCRIPTION
**Pull request by vadimkerr**
This PR restores an ability to notify users via email. Previously we used Sendgrid to send emails, this PR removes it and adds a new messaging backend that sends emails using SMTP.

- [x] Remove old code related to Sendgrid (as it's not needed anymore)
- [x] Remove old code related to email verification (as verification is not needed anymore)
- [x] Implement emails as a new notification backend using SMTP (based on https://github.com/grafana/oncall/pull/50)
- [x] Handle SMTP exceptions and add appropriate notification logs creation
- [x] Allow configuring SMTP using live settings
- [x] Review email limits
- [x] Add an ability to view & change messaging backend specific templates via public API 
- [x] Fix & add tests
- [x] Add SMTP configuration docs for OSS

Email example:
Screenshot 2022-10-13 at 16 13 28

Should resolve https://github.com/grafana/oncall/issues/347.

### CODE REVIEW

File: engine/apps/slack/scenarios/alertgroup_appearance.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def process_scenario(self, slack_user_identity, slack_team_identity, payload, ac
             blocks.append(block)
         blocks.append({"type": "divider"})
 
-        for notification_channel in ["slack", "web", "sms", "phone_call", "email", "telegram"]:
+        for notification_channel in ["slack", "web", "sms", "phone_call", "telegram"]:
```
Comment (matiasb) [inline engine/apps/slack/scenarios/alertgroup_appearance.py:93]: Can we use `NOTIFICATION_CHANNEL_OPTIONS` here? (and below)

**Review by matiasb — APPROVED** [review]
LGTM! Minor suggestions added

File: engine/apps/email/tasks.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        )
+        EmailMessage.objects.create(
+            represents_alert_group=alert_group,
+            notification_policy=notification_policy,
+            receiver=user,
+            exceeded_limit=True,
+        )
+        return
+
+    subject, html_message = build_subject_and_title(alert_group, emails_left)
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/email/tasks.py]: fixed in https://github.com/grafana/oncall/pull/621/commits/b6bef0363a04423b4974fc9a8298b9e59a2c9cec

File: engine/apps/base/messaging.py [hunk 3]
```diff
@@ -7,7 +7,9 @@ class BaseMessagingBackend:
     label = "The Backend"
     short_label = "Backend"
     available_for_use = False
+
     templater = None
+    template_fields = ["title", "message", "image_url"]
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/base/messaging.py]: done in https://github.com/grafana/oncall/pull/621/commits/b6bef0363a04423b4974fc9a8298b9e59a2c9cec

File: engine/apps/alerts/models/alert_receive_channel.py [hunk 4]
```diff
@@ -162,6 +161,8 @@ class AlertReceiveChannel(IntegrationOptionsMixin, MaintainableObject):
     web_message_template = models.TextField(null=True, default=None)
     web_image_url_template = models.TextField(null=True, default=None)
 
+    # email related fields are deprecated in favour of messaging backend based templates
+    # these templates are stored in the messaging_backends_templates field
     email_title_template = models.TextField(null=True, default=None)
     email_message_template = models.TextField(null=True, default=None)
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/alerts/models/alert_receive_channel.py]: Yes, these two fields will eventually be removed. I added more comments in https://github.com/grafana/oncall/pull/621/commits/05c514c168554691ec96334891f6e267b6d49f66 so it's super clear the fields are deprecated and shouldn't be used.

File: docs/sources/open-source/_index.md [hunk 5]
```diff
@@ -187,3 +187,12 @@ Grafana OnCall supports Twilio SMS and phone call notifications delivery. If you
 
 1. Set `GRAFANA_CLOUD_NOTIFICATIONS_ENABLED` as **False** to ensure the Grafana OSS <-> Cloud connector is disabled.
 1. From your **OnCall** environment, select **Env Variables** and configure all variables starting with `TWILIO_`. 
+
+## Email Setup
+Grafana OnCall is capable of sending emails using SMTP as a user notification step. To setup email notifications, populate the following env variables with your SMTP server credentials:
```
Comment (iskhakov) [inline docs/sources/open-source/_index.md:192]: may be write that step called "Notify by email" and which page is that

File: engine/apps/base/models/user_notification_policy_log_record.py [hunk 6]
```diff
@@ -49,14 +49,14 @@ class UserNotificationPolicyLogRecord(models.Model):
         ERROR_NOTIFICATION_PHONE_CALLS_LIMIT_EXCEEDED,
         ERROR_NOTIFICATION_PHONE_NUMBER_IS_NOT_VERIFIED,
         ERROR_NOTIFICATION_NOT_ABLE_TO_SEND_MAIL,
-        ERROR_NOTIFICATION_MAIL_LIMIT_EXCEEDED,
-        ERROR_NOTIFICATION_EMAIL_IS_NOT_VERIFIED,
+        ERROR_NOTIFICATION_MAIL_LIMIT_EXCEEDED,  # todo: manage backend specific limits in messaging backend
```
Comment (iskhakov) [inline engine/apps/base/models/user_notification_policy_log_record.py:52]: TODO

File: engine/apps/email/tasks.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+from apps.user_management.models import User
+from common.custom_celery_tasks import shared_dedicated_queue_retry_task
+
+MAX_RETRIES = 1 if settings.DEBUG else 10
+logger = get_task_logger(__name__)
+
+
+@shared_dedicated_queue_retry_task(autoretry_for=(Exception,), retry_backoff=True, max_retries=MAX_RETRIES)
+def notify_user_async(user_pk, alert_group_pk, notification_policy_pk):
+    from apps.base.models import UserNotificationPolicy, UserNotificationPolicyLogRecord
```
Comment (iskhakov) [inline engine/apps/email/tasks.py:21]: please comment why imported here

File: engine/apps/email/tasks.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    try:
+        user = User.objects.get(pk=user_pk)
+    except User.DoesNotExist:
```
Comment (iskhakov) [inline engine/apps/email/tasks.py:25]: check if we need to add logrecord
```
       UserNotificationPolicyLogRecord.objects.create(
             author=user,
             type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_FAILED,
```

File: engine/apps/email/tasks.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    try:
+        send_mail(subject, message, email_from, recipient_list, html_message=html_message, connection=connection)
+        EmailMessage.objects.create(
+            represents_alert_group=alert_group,
+            notification_policy=notification_policy,
+            receiver=user,
+            exceeded_limit=False,
+        )
+    except (gaierror, BadHeaderError) as e:
```
Comment (iskhakov) [inline engine/apps/email/tasks.py:85]: comments pls

File: engine/apps/email/templates/email_notification.html [hunk 10]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
```
Comment (iskhakov) [inline engine/apps/email/templates/email_notification.html:1]: may be use the same format as alerting? 
@raphael-batte can you please help Vadim
probably worth to do in the separate pr so we won't block this one

File: engine/apps/email/tests/test_notify_user.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    make_alert(alert_group=alert_group, raw_request_data=alert_receive_channel.config.example_payload)
+
+    notification_policy = make_user_notification_policy(
+        user,
+        UserNotificationPolicy.Step.NOTIFY,
+        notify_by=587,
+        important=False,
+    )
+
+    notify_user_async(user.pk, alert_group.pk, notification_policy.pk)
```
Comment (iskhakov) [inline engine/apps/email/tests/test_notify_user.py:43]: where are smtp settings defined for tests?

File: engine/config_integrations/grafana.py [hunk 12]
```diff
@@ -81,29 +81,6 @@
 
 phone_call_title = sms_title
 
-email_title = web_title
```
Comment (iskhakov) [inline engine/config_integrations/grafana.py:84]: please check other integrations in oncall-prvate enterprise integrations app

File: engine/settings/base.py [hunk 13]
```diff
@@ -569,6 +561,18 @@ class BrokerTypes:
...
+EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"
+EMAIL_HOST = os.getenv("EMAIL_HOST")
+EMAIL_HOST_USER = os.getenv("EMAIL_HOST_USER")
+EMAIL_HOST_PASSWORD = os.getenv("EMAIL_HOST_PASSWORD")
+EMAIL_PORT = getenv_integer("EMAIL_PORT", 587)
+EMAIL_USE_TLS = getenv_boolean("EMAIL_USE_TLS", True)
+DEFAULT_FROM_EMAIL = os.getenv("DEFAULT_FROM_EMAIL")
+
+if FEATURE_EMAIL_INTEGRATION_ENABLED:
+    EXTRA_MESSAGING_BACKENDS = [("apps.email.backend.EmailBackend", 587)]
```
Comment (iskhakov) [inline engine/settings/base.py:574]: I recommend to increment the id to match the current code style
however the idea with 587 is smart:)

**Comment by iskhakov** [comment]
Overall great work, some minor comments

**Review by iskhakov — APPROVED** [review]

File: engine/settings/base.py (hunk 13 above)
Comment (vadimkerr) [reply engine/settings/base.py]: Makes sense, I changed it to use 8 as backend id https://github.com/grafana/oncall/pull/621/commits/425d821cd83f08611092a90b06f2ba9b924d5ab2

File: engine/apps/email/tests/test_notify_user.py (hunk 11 above)
Comment (vadimkerr) [reply engine/apps/email/tests/test_notify_user.py:43]: Since `django.core.mail.backends.locmem.EmailBackend` is used as the django email backend, SMTP credentials are not needed and will be disregarded in case it's provided.

File: engine/apps/email/tasks.py (hunk 9 above)
Comment (vadimkerr) [reply engine/apps/email/tasks.py:86]: added in https://github.com/grafana/oncall/pull/621/commits/7151611d44f1c2283ae59796a4eb095d1ef824a4

File: engine/apps/email/tasks.py (hunk 7 above)
Comment (vadimkerr) [reply engine/apps/email/tasks.py:22]: Added in https://github.com/grafana/oncall/pull/621/commits/d2d15213b64b85dbb9c5ed070283752b2fa7f4e5

File: docs/sources/open-source/_index.md (hunk 5 above)
Comment (vadimkerr) [reply docs/sources/open-source/_index.md:192]: Added in https://github.com/grafana/oncall/pull/621/commits/e85ddf69b04338ccc6015232e243fd5cea984f9d

File: engine/apps/base/models/user_notification_policy_log_record.py (hunk 6 above)
Comment (vadimkerr) [reply engine/apps/base/models/user_notification_policy_log_record.py:52]: Lowercase "todo" is used in our codebase too (as well as "Todo"), not sure if it's really worth it
